renormalizer/mps/mp.py

```diff
@@ -428,17 +470,323 @@ def compress(self, temp_m_trunc=None, ret_s=False):
+                        ridx = imps + 1
+                else:
+                    assert False
+                logger.debug(f"optimize site: {cidx}")
+
+                ltensor = environ.GetLR(
+                    "L", lidx, self, mpo, itensor=None, method=lmethod, 
+                    mps_conj = mps.conj()
+                )
+                rtensor = environ.GetLR(
+                    "R", ridx, self, mpo, itensor=None, method=rmethod,
+                    mps_conj = mps.conj()
+                )
+
+                # get the quantum number pattern
+                qnbigl, qnbigr, qnmat = mps._get_big_qn(cidx)
+                
+                # center mo
+                cmo = [asxp(mpo[idx]) for idx in cidx] 
+                cms = [asxp(self[idx]) for idx in cidx]
+                if method == "1site":
+                    if cms[0].ndim == 3:
+                        # S-a   l-S
+                        #     d
+                        # O-b-O-f-O
+                        #     e
+                        # S-c   k-S
+
+                        path = [
+                            ([0, 1], "abc, cek -> abek"),
+                            ([2, 0], "abek, bdef -> akdf"),
+                            ([1, 0], "akdf, lfk -> adl"),
+                        ]
+                    elif cms[0].ndim == 4:
+                        # S-a   l-S
+                        #     d
+                        # O-b-O-f-O
+                        #     e
+                        # S-c   k-S
+                        #     g
+                        path = [
+                            ([0, 2], "abc, bdef -> acdef"),
+                            ([2, 0], "acdef, cegk -> adfgk"),
+                            ([1, 0], "adfgk, lfk -> adgl"),
+                        ]
+                    cout = multi_tensor_contract(
+                        path, ltensor, cms[0], cmo[0], rtensor
+                    )
+                else:
+                    if USE_GPU:
+                        oe_backend = "cupy"
+                    else:
+                        oe_backend = "numpy"
+                    if cms[0].ndim == 3:
+                        # S-a       l-S
+                        #     d   g
+                        # O-b-O-f-O-j-O
+                        #     e   h
+                        # S-c   m   k-S
+                    
+                        cout = oe.contract("abc, bdef, fghj, ljk, cem, mhk -> adgl",
+                                ltensor, cmo[0], cmo[1], rtensor, cms[0], cms[1],
+                                backend=oe_backend)
+                    elif cms[0].ndim == 4:
+                        # S-a       l-S
+                        #     d   g
+                        # O-b-O-f-O-j-O
+                        #     e   h
+                        # S-c   m   k-S
+                        #     n   p
+                        cout = oe.contract("abc, bdef, fghj, ljk, cenm, mhpk -> adngpl",
```

Comment:
In the one site branch (line 570), `multi_tensor_contract` is used, while here `oe.contract` is used. I wonder why this is the case because it'll be nice if we can use `oe.contract` without concerns in the future.

Reply:
In the 1site algorithm, the "optimal" contraction path is known and it is easy to write with `multi_tensor_contract` (`oe.contract` may have an overhead to find a contraction path).
In the 2site algorithm, the contraction path is more complicated and the "optimal" one is not clear (I'm lazy to compare all of them by hand.), so I use `oe.contract` (in my experience when cleaning up `cv`, `oe` performs better in complicated tensor contraction because the overhead to find an "optimal" path is minor.). 
But I haven't done any benchmark on the performance.

Reply:
So, as a rule of thumb, the best practice is to write simple contractions by hand while leaving complicated contractions to `oe.contract`. This makes sense to me. We can follow the guide and update it according to any benchmark results in the future.

Reply:
Agree.